lib/gitWorkflow.js

```diff
@@ -169,82 +184,68 @@ class GitWorkflow {
+        this.partiallyStagedFiles.map(file => this.execGit(['checkout', '--force', '--', file]))
       )
-      debug('Done adding files to index!')
-    }
+    } catch {} // eslint-disable-line no-empty
+  }
 
-    const modifiedFilesAfterAdd = await this.execGit(['status', '--porcelain'])
-    if (!modifiedFilesAfterAdd && !this.allowEmpty) {
+  /**
+   * Applies back task modifications, and unstaged changes hidden in the stash.
+   * In case of a merge-conflict retry with 3-way merge.
+   */
+  async applyModifications(ctx) {
+    debug('Adding task modifications to index...')
+    await Promise.all(
+      // stagedFileChunks includes staged files that lint-staged originally detected.
+      // Add only these files so any 3rd-party edits to other files won't be included in the commit.
+      this.stagedFileChunks.map(stagedFiles => this.execGit(['add', ...stagedFiles]))
+    )
+    debug('Done adding task modifications to index!')
+
+    const stagedFilesAfterAdd = await this.execGit(['diff', '--name-only', '--cached'])
+    if (!stagedFilesAfterAdd && !this.allowEmpty) {
       // Tasks reverted all staged changes and the commit would be empty
       // Throw error to stop commit unless `--allow-empty` was used
       ctx.gitApplyEmptyCommit = true
       handleError(new Error('Prevented an empty git commit!'), ctx)
     }
+  }
 
-    // Restore unstaged changes by applying the diff back. If it at first fails,
-    // this is probably because of conflicts between task modifications.
-    // 3-way merge usually fixes this, and in case it doesn't we should just give up and throw.
-    if (await this.hasPatch(PATCH_UNSTAGED)) {
-      debug('Restoring unstaged changes...')
-      const unstagedPatch = this.getHiddenFilepath(PATCH_UNSTAGED)
-      try {
-        await this.execGit([...GIT_APPLY_ARGS, unstagedPatch])
-      } catch (error) {
-        debug('Error while restoring changes:')
-        debug(error)
-        debug('Retrying with 3-way merge')
-
-        try {
-          // Retry with `--3way` if normal apply fails
-          await this.execGit([...GIT_APPLY_ARGS, '--3way', unstagedPatch])
-        } catch (error2) {
-          debug('Error while restoring unstaged changes using 3-way merge:')
-          debug(error2)
-          ctx.gitApplyModificationsError = true
-          handleError(
-            new Error('Unstaged changes could not be restored due to a merge conflict!'),
-            ctx
-          )
-        }
-      }
-      debug('Done restoring unstaged changes!')
-    }
-
-    // Restore untracked files by reading from the third commit associated with the backup stash
-    // See https://stackoverflow.com/a/52357762
+  /**
+   * Restore unstaged changes to partially changed files. If it at first fails,
+   * this is probably because of conflicts between new task modifications.
+   * 3-way merge usually fixes this, and in case it doesn't we should just give up and throw.
+   */
+  async restoreUnstagedChanges(ctx) {
```

Comment:
This new method applies back unstaged changes to files with staged changes, after running the tasks. The logic is the same as in v10, only the `--include=<file>` restricts it to only those files.